### PR TITLE
ABLASTR: Compute Phi

### PR DESCRIPTION
**Pull request by ax3l**
Expose the computation of the electro-static potential for relativistic beams via ABLASTR for ImpactX.

- [x] depends on merging #3024 first
- [x] depends on merging #3026 first
- [x] depends on merging #3167 first

### CODE REVIEW

File: Source/BoundaryConditions/PML.cpp [hunk 1]
```diff
@@ -17,6 +17,7 @@
 #include "Utils/WarpXAlgorithmSelection.H"
 #include "Utils/WarpXConst.H"
 #include "Utils/WarpXProfilerWrapper.H"
+#include "Utils/WarpXUtil.H"
```
Comment (ax3l) [inline Source/BoundaryConditions/PML.cpp]: fixed via #3138

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 2]
```diff
@@ -328,126 +333,70 @@ WarpX::computePhi (const amrex::Vector<std::unique_ptr<amrex::MultiFab> >& rho,
...
-#if defined(WARPX_DIM_RZ)
-        linop.setSigma({0._rt, 1._rt-beta_solver[1]*beta_solver[1]});
-#else
-        linop.setSigma({AMREX_D_DECL(
-            1._rt-beta_solver[0]*beta_solver[0],
-            1._rt-beta_solver[1]*beta_solver[1],
-            1._rt-beta_solver[2]*beta_solver[2])});
-#endif
-
+    std::optional<ElectrostaticSolver::EBCalcEfromPhiPerLevel> post_phi_calculation;
```
Comment (ax3l) [inline Source/FieldSolver/ElectrostaticSolver.cpp:282]: Oh no, I fear I screwed something up in this file in the last rebase

File: Source/ablastr/fields/PoissonSolver.H [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            const amrex::IntVect& refratio = rel_ref_ratio.value()[lev];
+            ba.coarsen(refratio);
+            const int ncomp = linop.getNComp();
+            amrex::MultiFab phi_cp(ba, phi[lev+1]->DistributionMap(), ncomp, 1);
+
+            // Copy from phi[lev] to phi_cp (in parallel)
+            const amrex::IntVect& ng = amrex::IntVect::TheUnitVector();
+            const amrex::Periodicity& crse_period = geom[lev].periodicity();
+            // TODO: move WarpXCommUtil.cpp over to ABLASTR
+            WarpXCommUtil::ParallelCopy(phi_cp, *phi[lev], 0, 0, 1, ng, ng, crse_period);
```
Comment (ax3l) [inline Source/ablastr/fields/PoissonSolver.H]: See #3167

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 4]
```diff
@@ -170,6 +174,7 @@ WarpX::AddSpaceChargeField (WarpXParticleContainer& pc)
     bool const reset = true;
     bool const do_rz_volume_scaling = true;
     pc.DepositCharge(rho, local, reset, do_rz_volume_scaling);
+    // ImpactX: add MPI comms to make deposition non-local
```
Comment (ax3l) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Cleanup:
```suggestion
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 5]
```diff
@@ -179,11 +184,11 @@ WarpX::AddSpaceChargeField (WarpXParticleContainer& pc)
     // Compute the potential phi, by solving the Poisson equation
     computePhi( rho, phi, beta, pc.self_fields_required_precision,
                 pc.self_fields_absolute_tolerance, pc.self_fields_max_iters,
-                pc.self_fields_verbosity );
+                pc.self_fields_verbosity ); // this
```
Comment (ax3l) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Cleanup:
```suggestion
                pc.self_fields_verbosity );
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 6]
```diff
@@ -76,7 +80,7 @@ WarpX::ComputeSpaceChargeField (bool const reset_fields)
             WarpXParticleContainer& species = mypc->GetParticleContainer(ispecies);
             if (species.initialize_self_fields ||
                 (do_electrostatic == ElectrostaticSolverAlgo::Relativistic)) {
-                AddSpaceChargeField(species);
+                AddSpaceChargeField(species); // this
```
Comment (ax3l) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Cleanup:
```suggestion
                AddSpaceChargeField(species);
```

File: Source/FieldSolver/ElectrostaticSolver.cpp [hunk 7]
```diff
@@ -179,11 +184,11 @@ WarpX::AddSpaceChargeField (WarpXParticleContainer& pc)
...
     computePhi( rho, phi, beta, pc.self_fields_required_precision,
                 pc.self_fields_absolute_tolerance, pc.self_fields_max_iters,
-                pc.self_fields_verbosity );
+                pc.self_fields_verbosity ); // this
 
     // Compute the corresponding electric and magnetic field, from the potential phi
-    computeE( Efield_fp, phi, beta );
-    computeB( Bfield_fp, phi, beta );
+    computeE( Efield_fp, phi, beta ); // modify for ImpactX to be nodal
+    computeB( Bfield_fp, phi, beta ); // modify for ImpactX to be nodal
```
Comment (ax3l) [inline Source/FieldSolver/ElectrostaticSolver.cpp]: Cleanup:
```suggestion
    computeE( Efield_fp, phi, beta );
    computeB( Bfield_fp, phi, beta );
```

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**